lib/http_v4_signature.c

```diff
@@ -0,0 +1,310 @@
+/***************************************************************************
+ *                                  _   _ ____  _
+ *  Project                     ___| | | |  _ \| |
+ *                             / __| | | | |_) | |
+ *                            | (__| |_| |  _ <| |___
+ *                             \___|\___/|_| \_\_____|
+ *
+ * Copyright (C) 1998 - 2020, Daniel Stenberg, <daniel@haxx.se>, et al.
+ *
+ * This software is licensed as described in the file COPYING, which
+ * you should have received as part of this distribution. The terms
+ * are also available at https://curl.haxx.se/docs/copyright.html.
+ *
+ * You may opt to use, copy, modify, merge, publish, distribute and/or sell
+ * copies of the Software, and permit persons to whom the Software is
+ * furnished to do so, under the terms of the COPYING file.
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
+ * KIND, either express or implied.
+ *
+ ***************************************************************************/
+
+#include "curl_setup.h"
+
+#include "urldata.h"
+#include "strcase.h"
+#include "vauth/vauth.h"
+#include "vauth/digest.h"
+#include "http_v4_signature.h"
+#include "curl_sha256.h"
+#include "transfer.h"
+
+/* The last 3 #include files should be in this order */
+#include "curl_printf.h"
+#include "curl_memory.h"
+#include "memdebug.h"
+#include "parsedate.h"
+
+#include <time.h>
+
+static CURLcode hmac_sha256(const unsigned char *key, unsigned int keylen,
+                            const unsigned char *data, unsigned int datalen,
+                            unsigned char *output)
+{
+  struct HMAC_context *ctxt = Curl_HMAC_init(Curl_HMAC_SHA256, key, keylen);
+
+  if(!ctxt)
+    return CURLE_OUT_OF_MEMORY;
+
+  /* Update the digest with the given challenge */
+  Curl_HMAC_update(ctxt, data, datalen);
+
+  /* Finalise the digest */
+  Curl_HMAC_final(ctxt, output);
+
+  return CURLE_OK;
+}
+
+#define HMAC_SHA256(k, kl, d, dl, o)                  \
+  if(hmac_sha256(k, kl, d, dl, o) != CURLE_OK) {      \
+    ret = CURLE_OUT_OF_MEMORY;                        \
+    goto free_all;                                    \
+  }
+
+#define PROVIDER_MAX_L 16
+
+CURLcode Curl_output_v4_signature(struct connectdata *conn, bool proxy)
+{
+  CURLcode ret = CURLE_OK;
+  char sk[45]; /* secret key is 40 chat long + 'OSC' + \0 */
+  struct Curl_easy *data = conn->data;
+  const char *customrequest = data->set.str[STRING_CUSTOMREQUEST];
+  const char *surl = strstr(data->set.str[STRING_SET_URL], "://") + 3;
```

Comment:
This is not safe code. What exactly is `surl`, shouldn't it rather use a decomposed URL part?